codersdk/toolsdk/toolsdk.go

```diff
@@ -13,47 +15,198 @@ import (
+	User              string            `json:"user"`
+}
+
+type CreateWorkspaceBuildArgs struct {
+	TemplateVersionID string `json:"template_version_id"`
+	Transition        string `json:"transition"`
+	WorkspaceID       string `json:"workspace_id"`
+}
+
+type DeleteTemplateArgs struct {
+	TemplateID string `json:"template_id"`
+}
+
+type GetTemplateVersionLogsArgs struct {
+	TemplateVersionID string `json:"template_version_id"`
+}
+
+type GetWorkspaceArgs struct {
+	WorkspaceID string `json:"workspace_id"`
+}
+
+type GetWorkspaceAgentLogsArgs struct {
+	WorkspaceAgentID string `json:"workspace_agent_id"`
+}
+
+type GetWorkspaceBuildLogsArgs struct {
+	WorkspaceBuildID string `json:"workspace_build_id"`
+}
+
+type ListWorkspacesArgs struct {
+	Owner string `json:"owner"`
+}
+
+type ListTemplateVersionParametersArgs struct {
+	TemplateVersionID string `json:"template_version_id"`
+}
+
+type UpdateTemplateActiveVersionArgs struct {
+	TemplateID        string `json:"template_id"`
+	TemplateVersionID string `json:"template_version_id"`
+}
+
+type UploadTarFileArgs struct {
+	Files map[string]string `json:"files"`
+}
+
+// WithRecover wraps a HandlerFunc to recover from panics and return an error.
+func WithRecover(h GenericHandlerFunc) GenericHandlerFunc {
+	return func(ctx context.Context, tb Deps, args json.RawMessage) (ret json.RawMessage, err error) {
+		defer func() {
+			if r := recover(); r != nil {
+				err = xerrors.Errorf("tool handler panic: %v", r)
+			}
+		}()
+		return h(ctx, tb, args)
+	}
+}
+
+// WithCleanContext wraps a HandlerFunc to provide it with a new context.
+// This ensures that no data is passed using context.Value.
+// If a deadline is set on the parent context, it will be passed to the child
+// context.
+func WithCleanContext(h GenericHandlerFunc) GenericHandlerFunc {
+	return func(parent context.Context, tb Deps, args json.RawMessage) (ret json.RawMessage, err error) {
+		child, childCancel := context.WithCancel(context.Background())
+		defer childCancel()
+		// Ensure that the child context has the same deadline as the parent
+		// context.
+		if deadline, ok := parent.Deadline(); ok {
+			deadlineCtx, deadlineCancel := context.WithDeadline(child, deadline)
+			defer deadlineCancel()
+			child = deadlineCtx
+		}
+		// Ensure that cancellation propagates from the parent context to the child context.
+		go func() {
+			select {
+			case <-child.Done():
+				return
+			case <-parent.Done():
+				childCancel()
+			}
+		}()
+		return h(child, tb, args)
+	}
+}
+
+// wrap wraps the provided GenericHandlerFunc with the provided middleware functions.
+func wrap(hf GenericHandlerFunc, mw ...func(GenericHandlerFunc) GenericHandlerFunc) GenericHandlerFunc {
+	for _, m := range mw {
+		hf = m(hf)
+	}
+	return hf
+}
+
 var (
 	// All is a list of all tools that can be used in the Coder CLI.
 	// When you add a new tool, be sure to include it here!
-	All = []Tool[any]{
-		CreateTemplateVersion.Generic(),
+	All = []GenericTool{
```

Comment:
It would be useful to have different lists depending on whether you have an agent client and slug.